### PR TITLE
Revert move of version.h to tf-psa-crypto library

### LINKED ISSUE #9620: Move back version and version_features to mbedtls
### Suggested enhancement
As decided as part of the review of #9236 we will not have version and version_features support to start with in TF-PSA-Crypto.
Thus move back `version.h` in `include/mbedtls` and revert to the generation of version_features.c in `library` by both make and CMake build system.

Move test_suite_version back to `tests/suites`.

### Justification
Mbed TLS needs this for the repo split.

### PR DESCRIPTION
**Pull request by Harry-Ramsey**
This commit moves version.h back into /include/mbedtls from tf-psa-crypto/drivers/builtin/include. This commit also changes the necessary build files and scripts to generate version_features.c

Fixes #9620

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** not required because: This change was for 4.0 but is now being reverted.
- [x] **development PR** provided.
- [x] **framework PR** not required.
- [x] **3.6 PR** not required because: this change was for 4.0.
- [x] **2.28 PR** not required because: this change was for 4.0
- **tests**  not required because: build configuration change.

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

Help make review efficient:
* Multiple simple commits
  - please structure your PR into a series of small commits, each of which does one thing
* Avoid force-push
  - please do not force-push to update your PR - just add new commit(s)
* See our [Guidelines for Contributors](https://mbed-tls.readthedocs.io/en/latest/reviews/review-for-contributors/) for more details about the review process.

### CODE REVIEW

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks quite good to me but we should move the version files to the TLS library (see my comments). Otherwise, please remove version_features.c from `tf-psa-crypto/drivers/builtin/src/.gitignore` and move `tf-psa-crypto/tests/suites/test_suite_version.data/function` to `tests/suites/` (thus from crypto test suites to TLS/x509 test suites).

File: library/CMakeLists.txt [hunk 1]
```diff
@@ -92,8 +92,8 @@ set(src_crypto
     ${TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_DIR}/sha3.c
     ${TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_DIR}/threading.c
     ${TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_DIR}/timing.c
-    ${TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_DIR}/version.c
-    ${TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_DIR}/version_features.c
+    version.c
+    version_features.c
```
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply library/CMakeLists.txt]: Thanks, resolved.

File: library/Makefile [hunk 2]
```diff
@@ -188,8 +188,8 @@ OBJS_CRYPTO= \
 	     $(TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_PATH)/sha3.o \
 	     $(TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_PATH)/threading.o \
 	     $(TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_PATH)/timing.o \
-	     $(TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_PATH)/version.o \
-	     $(TF_PSA_CRYPTO_DRIVERS_BUILTIN_SRC_PATH)/version_features.o \
+	     version.o \
+	     version_features.o \
```
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply library/Makefile]: Thanks, resolved.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Two minor things. Otherwise some CI components based on `config.py crypto` are falling when they check that TLS and x509 libraries are empty. As we move `version.o` and `version_features.o` to the TLS library we should reflect that in `config.py crypto`.
In `include_in_crypto` in `config.py` we should have something like:

```py
if name.startswith('MBEDTLS_X509_') or \                                    
   name.startswith('MBEDTLS_VERSION_') or \                                 
   name.startswith('MBEDTLS_SSL_') or \                                     
   name.startswith('MBEDTLS_KEY_EXCHANGE_'):                                
    return False                         
```

File: library/CMakeLists.txt [hunk 3]
```diff
@@ -109,6 +107,8 @@ set(src_x509
 )
 
 set(src_tls
+    version.c
```
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply library/CMakeLists.txt]: Thanks, addressed.

File: library/Makefile [hunk 4]
```diff
@@ -211,6 +209,8 @@ OBJS_X509= \
 	   # This line is intentionally left blank
 
 OBJS_TLS= \
+	  version.o \
```
[the remark this answers is not in the document]
Comment (Harry-Ramsey) [reply library/Makefile]: Thanks, addressed.

**Comment by ronald-cron-arm** [comment]
Two components failed so far as described in  #9652. This has been fixed by #9662. If you rebase these CI errors should disappear.

**Comment by ronald-cron-arm** [comment]
We have issues with psasim tests as we move the version modules to libmbedtls.a and for PSASIM server we are expecting them in libmbedcrypto.a. We can fix that by adding in all.sh after line 1019 (`scripts/config.py crypto_full`) `scripts/config.py set MBEDTLS_VERSION_C` and in `tests/psa-client-server/psasim/Makefile` changing line 8 to `SERVER_LIBS := -Lserver_libs -lmbedcrypto -lmbedtls`. When we will add "version" support in TF-PSA-Crypto we will be able to remove the dependency on `libmbedtls.a`. @tom-cosgrove-arm does this seem acceptable?

**Comment by tom-cosgrove-arm** [comment]
@ronald-cron-arm Yes, that would be fine.

However, as far as I am aware the Mbed TLS version support (`mbedtls_version_get_string_full()`) in PSASIM is only needed to print a sign-on message. Since that is a "nice to have", that bit could easily be removed if you wanted (maybe behind `#ifdef MBEDTLS_VERSION_C`). Ideally we would put that back when there is version support in TF-PSA-Crypto, but that's not essential

**Comment by ronald-cron-arm** [comment]
> @ronald-cron-arm Yes, that would be fine.
> 
> However, as far as I am aware the Mbed TLS version support (`mbedtls_version_get_string_full()`) in PSASIM is only needed to print a sign-on message. Since that is a "nice to have", that bit could easily be removed if you wanted (maybe behind `#ifdef MBEDTLS_VERSION_C`). Ideally we would put that back when there is version support in TF-PSA-Crypto, but that's not essential

Thanks. In "tests/psa-client-server/psasim/src/server.c", in psa_server_main():
```C
#if defined(MBEDTLS_VERSION_C)
    char mbedtls_version[18];             
    mbedtls_version_get_string_full(mbedtls_version);
    SERVER_PRINT("%s", mbedtls_version);
    #endif
```
seems to fix the CI issues. It is simpler than my proposal thus @Harry-Ramsey please try that.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks.

**Comment by ronald-cron-arm** [comment]
@minosgalanakis please have a look.

File: tf-psa-crypto/drivers/builtin/src/.gitignore [hunk 5]
```diff
@@ -1,4 +1,3 @@
 ###START_GENERATED_FILES###
 /error.c
-/version_features.c
```
Comment (minosgalanakis) [inline tf-psa-crypto/drivers/builtin/src/.gitignore:3]: Minor, the commit message is refering to the the version.c file while we are removing the _features.c.

No need to rework as its pretty obvious what is happening.

**Review by minosgalanakis — APPROVED** [review]
LGTM

**Comment by ronald-cron-arm** [comment]
The ABI-API-checking test component indicates that "version" symbols are removed from libmbedcrypto. That's expected.

**Pull request merged**